### PR TITLE
Update register model to support different register sizes

### PR DESCRIPTION
**Pull request by gembcior**

### CODE REVIEW

File: model/cpp/register_model.h [hunk 1]
```diff
@@ -31,107 +31,102 @@
...
-  static uint32_t read()
+  static SizeType read()
   {
-    volatile uint32_t* reg = reinterpret_cast<volatile uint32_t*>(address);
+    volatile SizeType* reg = reinterpret_cast<volatile SizeType*>(address);
     return (*reg >> position) & Mask;
   }
-
-  static_assert(position >= 0 && position <= 31, "Field position must be between 0 and 31");
-  static_assert(width >= 1 && width <= 32, "Field width must be between 1 and 32");
```
Comment (serweryn617) [inline model/cpp/register_model.h:125]: You could check something like `position >= 0 && position < sizeof(SizeType) * 8`

File: model/cpp/register_model.h [hunk 2]
```diff
@@ -168,27 +163,24 @@ class BitFieldModel
...
-
-  static_assert(position >= 0 && position <= 31, "BitFiled position must be between 0 and 31");
-  static_assert(width >= 1 && width <= 32, "BitFiled width must be between 1 and 32");
+  SizeType m_value;
 };
 
-template<uint32_t position>
-class BitFieldModel<position, 1U>
+template<typename SizeType, SizeType position, SizeType width>
+class BitFieldModel<SizeType, position, width, std::enable_if_t<width == 0>>
```
Comment (serweryn617) [inline model/cpp/register_model.h:178]: Pretty sure it should be `std::enable_if_t<width == 1>`

Comment (gembcior) [reply model/cpp/register_model.h]: Thanks, you are right.

**Comment by serweryn617** [comment]
D-RAL serial template uses the same int size for registers and their addresses which may lead to errors:

```cpp
  constexpr static uint8_t Address = 0x40000000 + 0x001B;
```

**Comment by serweryn617** [comment]
IMO a better approach is to use natural length integers for addresses and then allow .dral template processing to use custom fields. In such case an application specific template can be created with fields like address size added by the user when necessary.
I'm working on something similar on my fork, as I needed some additional metadata in register definitions, but it makes the entire project more flexible and helps in situations like this one.

**Pull request merged**